File: python/rotting-oranges.py
class Solution:
    def orangesRotting(self, grid: 'List[List[int]]') -> 'int':
        m, n=len(grid), len(grid[0])
        q=[]
        count, minutes=0, -1
        for i in range(m):
            for j in range(n):
                if grid[i][j]==2: q.append([i,j])
                elif grid[i][j]==1: count+=1
        if count==0: return 0
        
        while q:
            size=len(q)
            for i in range(size):
                f=q.pop(0)
                for nr, nc in [(f[0]-1, f[1]), (f[0],f[1]-1), (f[0]+1, f[1]), (f[0], f[1]+1)]:
                    if nr>=0 and nr<m and nc>=0 and nc<n and grid[nr][nc]==1:
                        grid[nr][nc]=2
                        count-=1
                        q.append([nr,nc])
            minutes+=1
            
        return minutes if count==0 else -1

File: cpp/rotting-oranges.cpp
994. Rotting Oranges
Easy

48

1

Favorite

Share
In a given grid, each cell can have one of three values:

the value 0 representing an empty cell;
the value 1 representing a fresh orange;
the value 2 representing a rotten orange.
Every minute, any fresh orange that is adjacent (4-directionally) to a rotten orange becomes rotten.

Return the minimum number of minutes that must elapse until no cell has a fresh orange.  If this is impossible, return -1 instead.

 

Example 1:



Input: [[2,1,1],[1,1,0],[0,1,1]]
Output: 4
Example 2:

Input: [[2,1,1],[0,1,1],[1,0,1]]
Output: -1
Explanation:  The orange in the bottom left corner (row 2, column 0) is never rotten, because rotting only happens 4-directionally.
Example 3:

Input: [[0,2]]
Output: 0
Explanation:  Since there are already no fresh oranges at minute 0, the answer is just 0.
 

Note:

1 <= grid.length <= 10
1 <= grid[0].length <= 10
grid[i][j] is only 0, 1, or 2.

class Solution {
public:
    int orangesRotting(vector<vector<int>>& grid) {
        int m=grid.size(), n=grid[0].size();
        queue<pair<int,int>> q;
        int count=0, minutes=-1;
        for (int i=0; i<m; ++i) {
            for (int j=0; j<n; ++j) {
                if (grid[i][j]==2) q.push({i,j});
                else if (grid[i][j]==1) count++;
            }
        }
        if (count==0) return 0;
        
        while (!q.empty()) {
            int size=q.size();
            for (int i=0; i<size; ++i) {
                auto f=q.front();
                q.pop();
                for (int k=0; k<4; ++k) {
                    int nr=f.first+delta[k][0];
                    int nc=f.second+delta[k][1];                    
                    if (nr>=0 && nr<m && nc>=0 && nc<n && grid[nr][nc]==1) {                        
                        grid[nr][nc]=2;
                        count--;
                        q.push({nr,nc});
                    }
                }
            }
            minutes++;
        }
        
        return count==0?minutes:-1;
    }
private:
    const int delta[4][2]={{-1,0}, {1,0}, {0,-1}, {0,1}};
};

